### PR TITLE
feat(ts): Heap class in TypeScript

### PR DESCRIPTION
**Pull request by Shyam-Chen**
If this pull request (PR) pertains to **Chinese-to-English translation**, please confirm that you have read the contribution guidelines and complete the checklist below:

- [ ] This PR represents the translation of a single, complete document, or contains only bug fixes.
- [ ] The translation accurately conveys the original meaning and intent of the Chinese version. If deviations exist, I have provided explanatory comments to clarify the reasons.

If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:

- [x] I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
- [x] I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
- [x] The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

可以直接在前端網頁上執行：

```html
<script type="module">
import { Heap } from 'https://esm.sh/heap-js@2.5.0';

/* 初始化堆 */
// 初始化小顶堆
const minHeap = new Heap();
// 初始化大顶堆
const maxHeap = new Heap(Heap.maxComparator);

/* 元素入堆 */
maxHeap.push(1);
maxHeap.push(3);
maxHeap.push(2);
maxHeap.push(5);
maxHeap.push(4);

/* 获取堆顶元素 */
let peek = maxHeap.peek(); // 5

/* 堆顶元素出堆 */
// 出堆元素会形成一个从大到小的序列
peek = maxHeap.poll(); // 5
peek = maxHeap.poll(); // 4
peek = maxHeap.poll(); // 3
peek = maxHeap.poll(); // 2
peek = maxHeap.poll(); // 1

/* 获取堆大小 */
const size = maxHeap.size();

/* 判断堆是否为空 */
const isEmpty = maxHeap.isEmpty();

/* 输入列表并建堆 */
minHeap.init([1, 3, 2, 5, 4]);
</script>
```

或是在 [Bun](https://bun.sh/) 上執行 TypeScript

或是改用 [Deno](https://deno.com/) (這沒有在此 PR)

```ts
import { BinaryHeap, ascend } from 'jsr:@std/data-structures';

/* 初始化堆 */
// 初始化小顶堆
const minHeap = new BinaryHeap<number>(ascend);
// 初始化大顶堆
const maxHeap = new BinaryHeap<number>();

/* 元素入堆 */
maxHeap.push(1);
maxHeap.push(3);
maxHeap.push(2);
maxHeap.push(5);
maxHeap.push(4);

/* 获取堆顶元素 */
let peek = maxHeap.peek(); // 5

/* 堆顶元素出堆 */
// 出堆元素会形成一个从大到小的序列
peek = maxHeap.pop(); // 5
peek = maxHeap.pop(); // 4
peek = maxHeap.pop(); // 3
peek = maxHeap.pop(); // 2
peek = maxHeap.pop(); // 1

/* 获取堆大小 */
const size = maxHeap.length;

/* 判断堆是否为空 */
const isEmpty = maxHeap.isEmpty();

/* 输入列表并建堆 */
const minHeap2 = BinaryHeap.from([1, 3, 2, 5, 4]);
```

---

此 PR 還額外移除了傳統 JavaScript 的寫法

### CODE REVIEW

File: docs/chapter_heap/heap.md [hunk 1]
```diff
@@ -304,13 +304,79 @@
 === "JS"
 
     ```javascript title="heap.js"
-    // JavaScript 未提供内置 Heap 类
+    import { Heap } from 'https://esm.sh/heap-js@2.5.0';
+
+    /* 初始化堆 */
+    // 初始化小顶堆
+    const minHeap = new Heap();
```
Comment (krahets) [inline docs/chapter_heap/heap.md:311]: Could we add the `heap.js` file which is runnable? So as `heap.ts`

Comment (Shyam-Chen) [reply docs/chapter_heap/heap.md:311]: `heap-js` is an npm library.

Comment (Shyam-Chen) [reply docs/chapter_heap/heap.md:311]: I have already added `heap.js` and `heap.ts`.

Comment (krahets) [reply docs/chapter_heap/heap.md:311]: Thanks! Sorry that I'm not familiar with JS and TS.
Waiting for the review by @justin-tse.

**Review by justin-tse — COMMENTED** [review]
Hello @Shyam-Chen, thank you for your contribution. I've reviewed the code and noticed that in this chapter, we're utilizing the built-in heap class. However, it's important to note that JavaScript does not have a built-in heap class, and the use of esm.sh may not be widely known within the JS community. Therefore, importing it may not be the best approach, as it could potentially confuse many TypeScript and JavaScript developers who are unfamiliar with it.

**Comment by krahets** [comment]
Thanks for the PR @Shyam-Chen ! We may not merge the changes because the library should match one of the cases:
1. the library is built-in to the programming language.
2. the library is introduced by LeetCode by default and becomes the standard library for algorithmic problems.

**Comment by Shyam-Chen** [comment]
`import { BinaryHeap } from 'jsr:@std/data-structures';` is part of Deno's standard library, or should we opt for this, or should we abandon this PR?

**Pull request closed without merging**